File: Python/data_structures/08_hashing_with_chains.py
import sys
from collections import namedtuple

Params = namedtuple("Params", ["prime", "x"])

PARAMETERS = Params(prime=1000000007, x=263)

class HashTable(object):
    def __init__(self, prime, x, cardinality):
        self._prime = prime
        self._x = x
        self._cardinality = cardinality
        self._table = [[] for _  in range(self._cardinality)]

    def _polynomial_hashing(self, string):
        value = 0
        for letter in reversed(string):
            value = (value * self._x + ord(letter)) % self._prime
        return value % self._cardinality

    def add(self, string):
        hash_value = self._polynomial_hashing(string)
        chain = self._table[hash_value]
        if string not in chain:
            chain.append(string)

    def delete(self, string):
        hash_value = self._polynomial_hashing(string)
        chain = self._table[hash_value]
        for ix, item in enumerate(chain):
            if item == string:
                del chain[ix]
                return

    def find(self, string):
        hash_value = self._polynomial_hashing(string)
        chain = self._table[hash_value]
        if string in chain:
            return "yes"
        return "no"

    def check(self, k):
        return " ".join([item for item in reversed(self._table[k])])


def read_queries(reader, nr_queries):
    return [tuple(next(reader).split()) for _ in range(nr_queries)]


def process_queries(queries, cardinality):
    hash_table = HashTable(*PARAMETERS, cardinality)
    results = []
    for operation, string in queries:
        if operation == "add":
            hash_table.add(string)
        elif operation == "del":
            hash_table.delete(string)
        elif operation == "find":
            results.append(hash_table.find(string))
        else:
            content = hash_table.check(int(string))
            results.append(content)
    return results


def main():
    reader = sys.stdin
    cardinality = int(next(reader))
    nr_queries = int(next(reader))
    queries = read_queries(reader, nr_queries)
    result = process_queries(queries, cardinality)
    for item in result:
        print(item)


if __name__ == "__main__":
    main()

File: Julia/data_structures/08-hashing-with-chains.jl
struct Query
    type::String
    value::Union{String, Nothing}
    k::Union{Int, Nothing}

    Query(type::String, value::String) = new(type, value, nothing)
    Query(type::String, k::Int) = new(type, nothing, k)
end


struct HashTable
    prime::Int
    x::Int
    cardinality::Int
    table::Vector{Vector{String}}

    function HashTable(prime::Int, x::Int, cardinality::Int)
        table = map(_ -> String[], 1:cardinality)
        new(prime, x, cardinality, table)
    end
end


function polynomialhash(hashtable::HashTable, s::String)
    foldr((char, acc) -> (acc*hashtable.x + Int(char)) % hashtable.prime, s; init=0) % hashtable.cardinality
end


function add!(hashtable::HashTable, s::String)
    hashvalue = polynomialhash(hashtable, s)
    chain = hashtable.table[hashvalue+1]
    if s ∉ chain
        push!(chain, s)
    end
end


function delete!(hashtable::HashTable, s::String)
    hashvalue = polynomialhash(hashtable, s)
    chain = hashtable.table[hashvalue+1]
    ix = findfirst(x -> x == s, chain)
    if ix != nothing
        deleteat!(chain, ix)
    end
end


function find(hashtable::HashTable, s::String)
    hashvalue = polynomialhash(hashtable, s)
    chain = hashtable.table[hashvalue+1]
    s ∈ chain ? "yes" : "no"
end


check(hashtable::HashTable, k::Int) = join(reverse(hashtable.table[k+1]), " ")


function readqueries(n)
    queries = Query[]
    for _ in 1:n
        a, b = map(String, split(readline()))
        query = a == "check" ? Query(a, parse(Int, b)) : Query(a, b)
        push!(queries, query)
    end
    queries
end


function processqueries(queries, cardinality, prime, x)
    hashtable = HashTable(prime, x, cardinality)
    results = String[]
    for query in queries
        operation = query.type
        if operation == "add"
            add!(hashtable, query.value)
        elseif operation == "del"
            delete!(hashtable, query.value)
        elseif operation == "find"
            push!(results, find(hashtable, query.value))
        else
            push!(results, check(hashtable, query.k))
        end
    end
    results
end


function main()
    cardinality = parse(Int, readline())
    nrqueries = parse(Int, readline())
    queries = readqueries(nrqueries)
    prime = 1000000007
    x = 263
    result = processqueries(queries, cardinality, prime, x)
    for item in result
        println(item)
    end
end


main()
